Fixes #178 Add Show-Object

tznind opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary
Adds the `out-ShowObject` command which presents the passed objects in a tree.  Below each object passed appear its public fields/properties.  

- Value, Null, String (or where an Exception was grenerated querying property) appear as leaf elements
- All other properties can be expanded as branches

![out-showobject](https://user-images.githubusercontent.com/31306100/210569074-c8380db1-9a5e-4f49-a509-221bc180c52b.gif)

<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->
This is my first time working with powershell commandlets so please be gentle.

I thought I would look into implementing #178 .  Let me know if this is not what you were expecting.  If its not wildly out.

I am happy to look into adding whatever else is required (e.g. filtering / highlighting etc).

tznind commented:
@microsoft-github-policy-service agree

claudio-salvio commented:
Hello @tznind 

First of all, thank you for your contribution

I would like to propose you to consider another name, `out-showobject` seems to me a bit strange.
Maybe, you could consider something like `Show-ObjectTree` or `Show-ObjectAsTree`.

Best regards

tznind commented:
> I would like to propose you to consider another name, out-showobject seems to me a bit strange.
Maybe, you could consider something like Show-ObjectTree or Show-ObjectAsTree.

Good point.  I've renamed it `Show-ObjectTree`.

I also added a StatusBar that shows original object count and selected leaf Type.  And I've made it expand `IEnumerable` values (except `string`) similarly to how VisualStudio watch window does:

![image](https://user-images.githubusercontent.com/31306100/212767194-63ddd099-8253-4c0c-8ada-7c78c55fa1d5.png)

MrFly72 commented:
I like this idea.
It helps debugging object values. That would even drive me to put it on a windows-box :-)

SteveL-MSFT commented:
@tznind This is really neat, are you still working on this?  I think you can do it iteratively when thinking about adding more features.

tig commented:
> @tznind This is really neat, are you still working on this? I think you can do it iteratively when thinking about adding more features.

I am now curious about how we can make ocgv and this work together seemlessly. By default selecting something in ocgv and having show-object launch will cause a bunch of screeen repaint. 

We really need non-full-screen Terminal.Gui apps... https://github.com/gui-cs/Terminal.Gui/issues/272

tznind commented:
> @tznind This is really neat, are you still working on this? I think you can do it iteratively when thinking about adding more features.

Yes! I'm excited to get back to this.  I agree an iterative approach would be good.  I've marked it ready for review.  But theres a lot that would be nice to add.

> I am now curious about how we can make ocgv and this work together seemlessly. By default selecting something in ocgv and having show-object launch will cause a bunch of screeen repaint.

Yes I'd love to explore this also.  Maybe a shortcut or something in ocgv to show a modal with the tree for currently selected object (or all objects)?

andschwa commented:
@tig, @tznind, and @SteveL-MSFT are we thinking this is good to review and potentially merge? I can't quite tell from the discussion above.

tznind commented:
> @tig, @tznind, and @SteveL-MSFT are we thinking this is good to review and potentially merge? I can't quite tell from the discussion above.

Its a yes from me.

In terms of iterative improvement of this feature, I have a PR in Terminal.Gui to support filtering in `TreeView<T>` component (https://github.com/gui-cs/Terminal.Gui/pull/2599).  

It is merged to develop and will be in the next library release.  Once that is out I can update this feature to have search.  But can do that on a subsequent PR.

tig commented:
If I can get T.Gui v1.11 out before Andrew gets to this we should include it. I'll try in next 24 hrs but no promises.

tig commented on src/Microsoft.PowerShell.ConsoleGuiTools/ShowObjectTreeCmdletCommand.cs:
```diff
@@ -0,0 +1,114 @@
+                baseObject is PSObject)
+            {
+                ErrorRecord error = new ErrorRecord(
+                    new FormatException("Invalid data type for Out-GridView"),
```
Should be "...for Show-Object".

tznind replied:
Fixed

tig commented on src/Microsoft.PowerShell.ConsoleGuiTools/ShowObjectTreeCmdletCommand.cs:
```diff
@@ -0,0 +1,114 @@
+    {
+        #region Properties
+        
+        private const string DataNotQualifiedForGridView = nameof(DataNotQualifiedForGridView);
```
Is this named correctly? Shouldn't it be "...ForShowObject"?

tig replied:
We should be consistent in parameters supported between OCGV and this. I think you should add:

* `-Title`
* `-Filter`
* `-MinUI`
* `-Focus` (if you want to tackle https://github.com/PowerShell/GraphicalTools/issues/127#issuecomment-1520433104)
* `-Search` (https://github.com/PowerShell/GraphicalTools/issues/127#issuecomment-1520433104)

tznind replied:
I have added in support for 

-Title
-Filter
-MinUI

I have made the filter logic and ux match console grid view (i.e. Regex support with error label).

I think for `-Focus` and `-Search` that if it is not in the base ocgv yet then its not much point implementing it in the tree view at this stage.

tznind replied:
For example you can now do:

```ps
($A = Get-ChildItem -Path $pshome -Recurse) | Show-ObjectTree -Title "Files in ps home" -MinUI -Filter manifest
```

![image](https://github.com/PowerShell/GraphicalTools/assets/31306100/8f1cfdf2-3270-40ee-8998-6ab6ecf30e9e)

tig commented on src/Microsoft.PowerShell.ConsoleGuiTools/ShowObjectView.cs:
```diff
@@ -0,0 +1,312 @@
+
+        public ShowObjectView(List<object> rootObjects)
+        {
+            Title = "Show-ObjectTree (Ctrl+Q to quit)";
```
I think the look & feel of OCGV and this should be the same. E.g. instead of the "Ctrl-Q to Quit" border title, add a statusbar. Also, OCGV uses Esc to quit as it's more standard.

tznind replied:
Now uses StatusBar and Escape

tig commented on src/Microsoft.PowerShell.ConsoleGuiTools/ShowObjectView.cs:
```diff
@@ -0,0 +1,312 @@
+                        Value = f.GetValue(parent);
+                    }
+                    else
+                        throw new NotSupportedException($"Unknown {nameof(MemberInfo)} Type");
```
Missing braces.

tznind replied:
Fixed

tig reviewed (changes requested):
I'm going to pull down and play with it, but here's my CR.

tig commented on src/Microsoft.PowerShell.ConsoleGuiTools/ShowObjectTreeCmdletCommand.cs:
```diff
@@ -0,0 +1,114 @@
+
+namespace OutGridView.Cmdlet
+{
+    [Cmdlet("Show", "ObjectTree")]
```
Not sure why, but lowercase is not working:

<img width="709" alt="image" src="https://github.com/PowerShell/GraphicalTools/assets/585482/d9ed7fa3-191b-4ac2-8272-457395c55faa">

Should we add an alias? e..g `sot`?

tznind replied:
You have a typo in your screenshot, you are missing hte second `t`

I have added an alias in b05aac7

tig reviewed (changes requested):
Usability issues:

-

tig commented:
I wanted to use `Show-ObjectTree` to act as an interactive version of `tree`. 

How would I do that?

tig commented on src/Microsoft.PowerShell.ConsoleGuiTools/Microsoft.PowerShell.ConsoleGuiTools.psd1:
```diff
@@ -69,7 +69,7 @@ NestedModules = @()
 
 # Cmdlets to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no cmdlets to export.
-CmdletsToExport = @( 'Out-ConsoleGridView' )
+CmdletsToExport = @( 'Out-ConsoleGridView', 'Show-ObjectTree' )
```
Add appropriate `Tags` below.

tznind replied:
Tags are now:

```
Tags = @('Console', 'Gui', 'Out-GridView', 'Out-ConsoleGridView', 'Show-ObjectTree', 'Terminal.Gui', 'gui.cs', 'MacOS', 'Windows', 'Linux', 'PSEdition_Core')
```

Only the command name is new, let me know if there are any other tags you think are now relevant.

tig reviewed (changes requested):
You need to add docs. See:

`src\Microsoft.PowerShell.ConsoleGuiTools\OutConsoleGridviewCmdletCommand.cs`

tznind commented:
> I wanted to use `Show-ObjectTree` to act as an interactive version of `tree`.
> 
> How would I do that?

So ` Get-ChildItem | Show-ObjectTree` probably makes the most sense for this use case.

This gives `DirectoryInfo` and `FileInfo` typed objects.  

The current behaviour of `Show-ObjectTree` for any object type is to show the public fields/properties of the objects passed.  In the case of `DirectoryInfo` this is `Length`, `IsReadOnly`, `LastAccessTime` etc.

We could detect when all the passed objects are IO objects and change the behaviour to enumerate children on expansion instead.

Optionally the user could pass a flag to prevent that behaviour?

tig commented:
@tznind - A bunch of the comments I added above have not been addressed in this PR.

- docs
- lowercase `show-object` isn't working
- inconsistent UI with ocgv
- ...

Will you be able to tackle these sometime soon?

tznind commented:
Yes sorry has fallen down my list a bit but can look again at it.

ALthough with the new TreeTable data source in v2 I have been wondering if
just adding a --tree flag to existing command would be easier.

It would allow viewing and selecting child elements in line.  For example
pipe some folders to ocgv then expand one and select some files within as
the ogcv command output.

On Wed, 19 Jul 2023, 15:45 Tig, ***@***.***> wrote:

> @tznind <https://github.com/tznind> - A bunch of the comments I added
> above have not been addressed in this PR.
>
>    - docs
>    - lowercase show-object isn't working
>    - inconsistent UI with ocgv
>    - ...
>
> Will you be able to tackle these sometime soon?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PowerShell/GraphicalTools/pull/179#issuecomment-1642229729>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHO3C5FGMXTEYTTZIXFI4VDXQ7XJHANCNFSM6AAAAAATQ2JIDY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tznind commented:
> I wanted to use `Show-ObjectTree` to act as an interactive version of `tree`.
> 
> How would I do that?

The default 'child getter' is to enumerate the properties on the PSObject and express as `<prop name>:<prop value>`.  This works well for all objects I have tried it with.

I have added a 'special cases' method called `GetExtraChildren` which allows adding additional tree elements for specific types of objects.  So now DirectoryInfo objects also show child files/directories.

![image](https://github.com/PowerShell/GraphicalTools/assets/31306100/b9944344-721d-4c79-86df-a3d570e623db)

andyleejordan commented:
@tig are you happy enough with this for me to do a review and potential merge now?

StartAutomating commented:
@tznind @tig really cool work!

Can't wait to figure out how it will integrate with [Write-FormatTreeView](https://github.com/StartAutomating/EZOut/blob/master/Commands/Format.PS1XML/Write-FormatTreeView.ps1) so that any object can be formatted as a tree!

andyleejordan commented:
> @tznind - A bunch of the comments I added above have not been addressed in this PR.

@tig it looks like this has been updated since you last looked. I'd prefer to merge this before the next release! Do you agree?

tig commented:
> > @tznind - A bunch of the comments I added above have not been addressed in this PR.
> 
> @tig it looks like this has been updated since you last looked. I'd prefer to merge this before the next release! Do you agree?

@tznind can you add support for `-Debug` and `-UseNetConsole` as I did to ocgv so we remain consistent?

tznind commented:
> > > @tznind - A bunch of the comments I added above have not been addressed in this PR.
> > 
> > 
> > @tig it looks like this has been updated since you last looked. I'd prefer to merge this before the next release! Do you agree?
> 
> @tznind can you add support for `-Debug` and `-UseNetConsole` as I did to ocgv so we remain consistent?

I have added these settings in [3c654e9](https://github.com/PowerShell/GraphicalTools/pull/179/commits/3c654e9b8de6d5a78803ad74c6bbf51a63f92eb9)

![image](https://github.com/PowerShell/GraphicalTools/assets/31306100/53728475-2114-4496-b455-3dd80128d7e1)

andyleejordan reviewed (approved):
This looks good to me! I'm going to merge it and then run it through an automated formatter because I don't believe in requesting just formatting / whitespace changes. I'm not super familiar with the code but I didn't spot anything obvious. Let's get it in!
